File: visual-search-python/imagesegmenter.py
import os, sys
from array import array
import cv2
import numpy as np
import json
import cv
#import pygame
def SplitIntoLines(bookimgdirpath):
    img = cv2.imread( bookimgdirpath,cv2.CV_LOAD_IMAGE_GRAYSCALE )
    ret, otsu =cv2.threshold(img,0,255,cv2.THRESH_BINARY+cv2.THRESH_OTSU)
    cv2.imwrite("binary-thresholded.png",otsu)
    
    m_white_row_threshold = 0
    m_white_row_fill_factor = 1.0
    m_white_col_threshold = 0
    m_white_col_fill_factor = 1.0
    m_MIN_INTER_WORD_GAP = 5
    
    m_lineBoundaries=[]
    m_lineBoundaries1=[]
    rows,cols = img.shape
    rowSums = []
    #rowSums = np.array([])
    rowSums = rows
    rowSum =0
    rowSums1 = []
    data = []
    #print rowSums1
    m_white_row_threshold=0 
    for row in range(rows):
        rowSum =0;
       
        for col in range(cols):
            rowSum += otsu[row][col]
            
        #print rowSum
        
        rowSums=rowSum/255
        #print rowSums
        for i in np.nditer(rowSums):
            #print i
            rowSums1.append(i)
    m_white_row_threshold = cols * m_white_row_fill_factor
    #print m_white_row_threshold 
    #out3 = cv2.zeros(rows,cols,cv2.CV_8UC3)
    size = rows, cols,3
    #out3 = np.zeros(size,dtype=cv2.uint8)
    out3=cv.CreateImage((rows,cols),cv.CV_8UC3,1)    
    #print rowSums
    channels =[]
    channels.append(otsu)
    channels.append(otsu)
    channels.append(otsu)
    out3=cv2.merge(channels)
    #cv2.imwrite('linerect.png',out3)
    lineTop=0
    lineBottom=0
    #lineB =[]
    arr = []
    for j in range (len(rowSums1)):
        #print "prev:",(rowSums1[j-1]),":","curr:",(rowSums1[j])
        if( rowSums1[j-1] >=m_white_row_threshold and rowSums1[j] <= m_white_row_threshold ):
            lineTop = j
        elif(rowSums1[j-1] <= m_white_row_threshold and rowSums1[j] >=  m_white_row_threshold ):
               lineBottom = j
               lineB = [lineTop,lineBottom]
               #data.append(lineTop)
               #data.append(lineBottom)
               #print lineB
               
               
               m_lineBoundaries.append(lineB[0])
               m_lineBoundaries1.append(lineB[1])
               #print m_lineBoundaries
               p1=(0,lineTop)
               p2=(cols-1,lineBottom)
               cv2.rectangle(out3,p1,p2,cv.CV_RGB(255,0,255),1,cv.CV_AA,0)
               SplitLinesIntoWords(otsu,lineB[0],lineB[1],cols,rows)                
    cv2.imwrite('result.png',out3)
   
    #print m_lineBoundaries
    #print lineB rowSums=rowSum/255
    #SplitLinesIntoWordsnew(otsu,m_lineBoundaries,m_lineBoundaries1,cols,rows)
    #SplitLinesIntoWords(otsu,m_lineBoundaries,m_lineBoundaries1,cols,rows) 
    
     
#def SplitLinesIntoWordsnew(img,m_lineBoundaries,m_lineBoundaries1,cols,rows): 
    #print m_lineBoundaries

    #for i in range(len(m_lineBoundaries)):
        
        #print m_lineBoundaries[i] 
   
            
    
    
    


    #cv2.imshow('result',img)
    #cv2.waitKey(2000)

def SplitLine(img,wordBoundaries,cols,rows):
    #cv2.imshow('img',img)
    #cv2.waitKey(0)
   
    col_sums = []
    col_sums1 = []
    col_sums = cols
    
    #col_sum = 0
    #for col in range(cols):
       # col_sum = 0
        #for row in range(rows):
            #col_sum += img[row][col]
            
        #col_sums = col_sum/255
        #print col_sums
        #for i in np.nditer( col_sums ):
            #col_sums1.append( i )
        #print col_sums1

        
def SplitLinesIntoWords(img,lineB1,lineB2,cols,rows):
    m_lineBoundaries=[lineB1,lineB2]

    currLineTop=lineB1
    currLineBottom=lineB2
    #print currLineTop
    #print m_lineBoundaries
    #currLineBoundaries=[][]
    #len1= len(m_lineBoundaries)
    #len2=len(m_lineBoundaries1)
    #m_lineB = m_lineBoundaries
    #print m_lineB
    #currLineTop1 =[]
    #currLineBottom1 = []
    #for i in range(len(m_lineBoundaries)):
        #currLineTop = m_lineBoundaries[i]
        #print "curlinetop:",currLineTop
        #currLineTop1.append(currLineTop)
    #print currLineTop1
    #for i in range(len(m_lineBoundaries)):
        #for j in range(len(m_lineBoundaries1)):
            #print m_lineBoundaries[i],";",m_lineBoundaries1[j]
            #currLineTop = m_lineBoundaries[i]
            #currLineBottom=m_lineBoundaries1[j]
            #print currLineBottom
     
    lineImg=cv.CreateImage((rows,cols),cv.CV_8UC3,1)
    channels =[]
    channels.append(img)
    channels.append(img)
    channels.append(img)
    lineImg=cv2.merge(channels)
    #cv2.rectangle(out3,p1,p2,cv.CV_RGB(255,0,255),1,cv.CV_AA,0)
    #cv2.rectangle(lineImg,0,currLineTop,cols,currLineBottom-currLineTop+1)
    #lineImg=cv2.rectangle(img,0,currLineTop,cols,currLineBottom-currLineTop+1)
    #print currLineTop,";",cols,";",currLineBottom-currLineTop+1
    p1=(0,currLineTop)
    p2=(cols,currLineBottom-currLineTop+1)

    cv2.rectangle(lineImg,p1,p2,cv.CV_RGB(255,0,255),1,cv.CV_AA)
    wordBoundaries = []
    SplitLine(lineImg,wordBoundaries,cols,rows)

    #cv2.imwrite('demo.png',lineImg)





def SplitIntoWords(bookimgdirpath,segmentedimgpath,outfile):
    #print bookimgdirpath,segmentedimgpath,outfile
    SplitIntoLines(bookimgdirpath)
    #SplitLinesIntoWords(bookimgdirpath);
    






print "initialized image Segementer..."
bookimgdata = []
with open('segment.json') as f :
     d=json.load( f )

bookimgdirpath=d['book'] ['imagedirpath']
bookdirs=os.listdir( bookimgdirpath)

for bookfilename in bookdirs:
        #print bookfilename
    bookimgdata.append(bookfilename)

segmentedimgpath=d['page'] ['segmentedimgpath']  
 
for i in range(len(bookimgdata)):
    #print bookimgdata[i]
    #print (bookimgdirpath+bookimgdata[i])
    SplitIntoWords(bookimgdirpath+bookimgdata[i],segmentedimgpath,"segmented_res"+bookimgdata[i])



   


















    

File: cpp/ImageSegmenter.orig/ImageSegmenter.cpp
#include "ImageSegmenter.h"
#include <opencv2/imgproc/im gproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/core/core.hpp>
#include <iostream>

ImageSegmenter::ImageSegmenter(const string& config_file_path)
{	
	m_white_row_fill_factor = 0.99;	
	m_white_col_fill_factor = 1.0;

	m_min_inter_word_gap = 10; // A gap larger than this means we have a word

	m_whiten_band_height = 15;
	m_whiten_band_width = 15;

	m_do_dilation = false;
	m_line_element_length = 3;

	m_roi_tlx = m_roi_tly = m_roi_brx = m_roi_bry = 0;

	m_cfgParser.ReadParamFile(config_file_path);
	if (m_cfgParser.good())
		ReadConfigParams();
}

void ImageSegmenter::SaveOutput(const string& output_type)
{	

    Mat out = Mat::zeros(m_docImage.rows,m_docImage.cols,CV_8UC3);
	vector<Mat> channels;
	channels.push_back(m_docImage);
	channels.push_back(m_docImage);
	channels.push_back(m_docImage);
	merge(channels,out);

	ofstream ofs;
	ofs.open(m_output_file_path, ios::out);


	if (!ofs) {
		cerr << "Could not open output file " << m_output_file_path << " for writing\n";
		return;
	}

	vector<string> lines;
	if (output_type == "LINES") {
		for (unsigned int i = 0; i < m_lineBoundaries.size(); i++) {
			pair<unsigned int, unsigned int> lineB = m_lineBoundaries.at(i);
			stringstream ss;
			ss << 0 + m_roi_tlx << " " << lineB.first + m_roi_tly << " " << m_docImage.cols - 1 << " " << lineB.second + m_roi_tly;
			lines.push_back(ss.str());
			ss.str("");			
		}			
	}
	else if (output_type == "WORDS") {
		for (unsigned int i = 0; i < m_wordBoundaries.size(); i++) {
			vector< pair<unsigned int, unsigned int> > wordBVec = m_wordBoundaries.at(i);
			pair<unsigned int, unsigned int> lineB = m_lineBoundaries.at(i);
			for (unsigned int j = 0; j < wordBVec.size(); j++) {
				pair<unsigned int, unsigned int> wordB = wordBVec.at(j);
				stringstream ss;
				ss << wordB.first + m_roi_tlx << " " << lineB.first + m_roi_tly << " " << wordB.second + m_roi_tlx << " " << lineB.second + m_roi_tly;
				lines.push_back(ss.str());
				ss.str("");
                               Point p1(wordB.first + m_roi_tlx,lineB.first + m_roi_tly);
                               Point p2(wordB.second + m_roi_tlx,lineB.second + m_roi_tly);
                               cv::rectangle(out,p1,p2,Scalar(255,0,255),1,CV_AA,0);
                               //imshow("Segmented Image", out);		
		               imwrite("Segment.jpg",out);
			}
		}
                               waitKey(0);
	}

	// http://stackoverflow.com/a/16330323/482389
	// remove the line begin with character '\n'   
	std::remove_if(lines.begin(), lines.end(), [](string &line) { return line[0] == '\n'; });
	// append newline to every line except last line  
	std::for_each(lines.begin(), lines.end() - 1, [](string &line) { line += '\n'; });
	// write lines to new file
	for (unsigned int i = 0; i < lines.size(); i++)
		ofs << lines.at(i);
	ofs.close();

}

bool ImageSegmenter::Process()
{
	// Check if we are working on full image or ROI
	if (m_processing_roi.length() > 0) { // ROI
		Mat inputImage = imread(m_document_image_path.c_str(), CV_LOAD_IMAGE_GRAYSCALE); // converts to grayscale if required
		stringstream ss(m_processing_roi);
		
		ss >> m_roi_tlx >> m_roi_tly >> m_roi_brx >> m_roi_bry;
		
		Rect roi(m_roi_tlx, m_roi_tly, m_roi_brx - m_roi_tlx + 1, m_roi_bry - m_roi_tly + 1);
		m_docImage = inputImage(roi);
		//imshow("roi", m_docImage);
		//waitKey();
	}
	else {
		// Read input image
		m_docImage = imread(m_document_image_path.c_str(), CV_LOAD_IMAGE_GRAYSCALE); // converts to grayscale if required
	}
	


	PreprocessImage(m_docImage);

	SplitIntoLines(m_docImage);	

	if (m_output_level == "LINES") {
		SaveOutput(m_output_level);
		return true;
	}
	else {
		SplitLinesIntoWords(m_docImage);
		SaveOutput(m_output_level);
	}
			
	return true;
}



void ImageSegmenter::PreprocessImage(Mat& img)
{
	// Convert input grayscale image to binary image
	threshold(img, m_docImage, 0, 255, CV_THRESH_OTSU);
	
	// 'White'-en a band around the image boundaries
	img(Rect(0, 0, img.cols, m_whiten_band_height)) = Scalar(255); 
	img(Rect(0, img.rows - m_whiten_band_height, img.cols, m_whiten_band_height)) = Scalar(255);
	img(Rect(0, 0, m_whiten_band_width, img.rows)) = Scalar(255);
	img(Rect(img.cols - m_whiten_band_width,0,m_whiten_band_width, img.rows)) = Scalar(255);
}


Mat ImageSegmenter::GetLineElement()
{
	Mat M = Mat::zeros(m_line_element_length, m_line_element_length, CV_8UC1);
	M(Rect(0, m_line_element_length / 2, m_line_element_length, 1)) = Scalar(1);
	return M.clone();
}

void ImageSegmenter::SplitIntoLines(const Mat& img)
{
	Mat processImg, dilated_img;
	if (m_do_dilation) {		
		Mat element = GetLineElement();
		dilated_img = img.clone();
		dilate(img, dilated_img, element);
		//imshow("before",img);
		//imshow("after", dilated_img);		
		//waitKey();
		processImg = dilated_img;
	}
	else
		processImg = img;
	
	// Form row histogram of sum-across-columns
	vector<unsigned int> rowSums(processImg.rows,0);
	unsigned int rowSum =0;
	for(unsigned int row = 0; row < processImg.rows; ++row) {
		rowSum =0;
		for(unsigned int col = 0; col < processImg.cols; ++col)
			rowSum += processImg.at<unsigned char>(row,col);
		rowSums.at(row) = rowSum/255;
		//cout << "Row = " << row << " RowSum = " << rowSums.at(row) << endl;
	}

	// If an entry in the row histogram exceeds a threshold,
	// it is considered a white line. Otherwise, the entry
	// implies it is part of a text line in the document. We scan the row
	// histogram from top to bottom looking for transitions from white line
	// to text and vice-versa.
	unsigned int white_row_threshold = static_cast<unsigned int>(static_cast<double>(processImg.cols) * m_white_row_fill_factor);
	unsigned int lineTop=0,lineBottom=0;
	
	for(unsigned int i = 1 ; i < rowSums.size() ;i++) {
		//cout << "Rowsums: prev[" <<  rowSums.at(i-1) << "] curr[" << rowSums.at(i) << "]" << endl;
		if ( rowSums.at(i-1) >= white_row_threshold && rowSums.at(i) <= white_row_threshold ) 
			lineTop = i;
		else if ( rowSums.at(i-1) <= white_row_threshold && rowSums.at(i) >=  white_row_threshold ) {
			lineBottom = i;

		pair<unsigned int,unsigned int> lineB;
		lineB.first = lineTop;
		lineB.second = lineBottom;
		m_lineBoundaries.push_back(lineB);
		//cout << "Line: " << lineTop << " " << lineBottom << endl;		
		}
	}
	
}		

void ImageSegmenter::SplitLinesIntoWords(const Mat& img)
{
	
	// Scan the line boundaries array and for each entry
	// split the line into words
	//cout << "Lines size=" << m_lineBoundaries.size() << endl;
	for(unsigned int i = 0 ; i < m_lineBoundaries.size() ; i++) {
		pair<unsigned int,unsigned int> currLineBoundaries = m_lineBoundaries.at(i);
		unsigned int currLineTop = currLineBoundaries.first;
		unsigned int currLineBottom = currLineBoundaries.second;
		//cout << "Line: " << currLineTop << " " << currLineBottom << endl;
		//cout << "Rect [0," << currLineTop << "," << m_docImage.cols << "," << currLineBottom-currLineTop+1 << "]" << endl; 
		Mat lineImg(m_docImage,cv::Rect(0,currLineTop,m_docImage.cols,currLineBottom-currLineTop+1));

		vector< pair<unsigned int,unsigned int> > wordBoundaries;
		SplitLine(lineImg,wordBoundaries);	
		m_wordBoundaries.push_back(wordBoundaries);
	}
		
}
